Display hostnames in reports when available

dav3r opened the pull request:
<!-- GitHub renders PRs such that soft line breaks are treated as hard
line breaks.  In order to make this template render as expected we
therefore have to avoid soft breaks and therefore will offend
markdownlint with long lines.  This is the reason for the markdownline
disable directive just below.

For more details see:
https://github.com/github/markup/issues/1050#issuecomment-294654762 -->
<!-- markdownlint-disable MD013 -->
# <!-- Use the title to describe PR changes in the imperative mood --> #

## 🗣 Description ##

This PR updates various CyHy reports (customer, daily notification, and Cyber Exposure scorecard) to display hostnames in the tables, appendices, and CSV attachments if the tickets associated with the report contain hostnames.  There are also a couple of bug fixes that I implemented along the way.

If an entity has no hostnames in their inventory, there will not be any significant changes to their customer report after this PR is deployed.  However, there are some minor changes in this PR that affect **all reports**, regardless of whether or not the entity has any hostnames to their inventory:
* In the `hosts.csv` attachment, the hostname column is now first (previously it was last) and the rows are now sorted by hostname, rather than ip_int.
* The `scope.csv` attachment is renamed to `scope-ip.csv`.  This is to reduce any confusion between that file and the `scope-hostname.csv` file that will be attached if the entity has hostnames in their inventory.

Additionally, thanks to the changes in https://github.com/cisagov/ncats-webd/pull/40, the following files will also include hostnames when they are generated as part of the weekly reporting process:
* `cybex_open_tickets_urgent.csv`
* `cybex_closed_tickets_urgent.csv`

<!-- Describe the "what" of your changes in detail. -->
<!-- To avoid scope creep, limit changes to a single goal. -->

## 💭 Motivation and context ##

This is a feature that has long been requested and now we are finally able to implement it.

This PR will resolve (partially) https://github.com/cisagov/cyhy-system/issues/70, in conjunction with:
* https://github.com/cisagov/cyhy-core/pull/101
* https://github.com/cisagov/cyhy-commander/pull/14
* https://github.com/cisagov/ansible-role-cyhy-commander/pull/88
* https://github.com/cisagov/cyhy_amis/pull/859
* https://github.com/cisagov/ncats-webd/pull/40

<!-- Why is this change required? -->
<!-- What problem does this change solve? How did you solve it? -->
<!-- Mention any related issue(s) here using appropriate keywords such -->
<!-- as "closes" or "resolves" to auto-close them on merge. -->

## 🧪 Testing ##

* [x] Deployed a database instance with hostname scanning support in a test environment
* [x] Imported request doc JSON for several test entities containing a mix of IP addresses, hostnames that resolve to an IP address owned by the test entity, and hostnames that resolve to IPs that resolve to an IP address not owned by any test entities.
* [x] Waited for test scans to complete and confirmed a variety of vulnerability findings were captured in the test database.
* [x] Generated test snapshots
* [x] Generated test reports and confirmed that hostnames are present throughout the reports when they should be and not present for test entities that have no hostnames in their tickets.
* [x] Generated reports for a variety of production entities (no logging to production DB) and confirmed that the reports are unchanged (since no production entities have hostnames yet)
* [x] Generated a Cyber Exposure Scorecard using the updated code from this PR and confirmed that it worked as expected


<!-- How did you test your changes? How could someone else test this PR? -->
<!-- Include details of your testing environment, and the tests you ran to -->
<!-- see how your change affects other areas of the code, etc. -->

<!--
## 📷 Screenshots (if appropriate) ##

Uncomment this section if a screenshot is needed.

-->

## ✅ Pre-approval checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - *eschew scope creep!*
- [x] All relevant type-of-change labels have been added.
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).

## ✅ Post-merge checklist ##

<!-- Remove any of the following that do not apply. -->

- [x] Create a release (necessary if and only if the version was bumped).
- [x] Build fresh production AMI
- [x] Deploy updated reporter instance to production
- [x] Verify daily notifications generate normally (entities without hostnames)
- [ ] Verify daily notifications generate normally (entities with hostnames)
- [x] Verify weekly reports generate normally (entities without hostnames)
- [x] Verify weekly reports generate normally (entities with hostnames)
- [x] Verify Cyber Exposure scorecard generates normally

dav3r committed:
Bump version from 1.0.0 to 1.1.0

dav3r committed:
Require cyhy-core 1.0.0, which supports hostnames in the database

dav3r committed:
Include hostnames (if available) in notification attachments

dav3r committed:
Redact hostnames in anonymized notifications

dav3r committed:
Display hostnames (if available) in notification tables

dav3r committed:
Switch to landscape orientation for notifications

The extra page width is needed to accomodate the additional hostname column in the "Urgent Vulnerabilities" table.

dav3r committed:
Make hostname column wider in risky services table for notifications

dav3r committed:
Set has_hostnames variable if any tickets have a hostname included

dav3r committed:
Update report queries to include HostDocs where the specified owners own a hostname, even if they don't own the HostDoc itself

This is needed now that HostDocs can be owned by the default CYHY owner, but contain hostnames owned by other CyHy entities.

dav3r committed:
Update findings attachment to include hostnames

Also, simplify logic for adding/removing various fields based on certain conditions.

dav3r committed:
Update mitigated vulns attachment to include hostnames

Also, simplify logic for adding/removing various fields based on certain conditions.

dav3r committed:
Update recently detected vulns attachment to include hostnames

Also, simplify logic for adding/removing various fields based on certain conditions.

dav3r committed:
Add hostnames (if any) to data used in the services attachment

Only include hostnames owned by the same owner(s) present in the snapshot for this report.

dav3r committed:
Add an additional flag to indicate if the hosts have any relevant hostnames

Also, rename a previous flag to indicate if the tickets have any relevant hostnames

dav3r committed:
Update services attachment to include hostnames

Also, simplify logic for adding/removing various fields based on certain conditions.

dav3r committed:
Update risky services attachment to include hostnames

Also, simplify logic for adding/removing various fields based on certain conditions.

dav3r committed:
Update false positives attachment to include hostnames

Also, simplify logic for adding/removing various fields based on certain conditions.

dav3r committed:
Anonymize hostname in all tickets for anonymized reports

dav3r committed:
Bugfix: anonymize risky services if requested

dav3r committed:
Simplify logic used to generate services attachment

This can be done now that cyhy-commander creates port_scan docs for each hostname/owner combo in a HostDoc, so there is no longer any need to figure out the relevant hostname(s) at reporting time.

Note that this commit also makes hostname the primary sort key for the services attachment (followed by ip_int, then port).

dav3r committed:
Update hosts attachment to include HostDoc hostnames

Also:
* Anonymize hostnames if requested
* Simplify logic for adding/removing various fields based on certain conditions.

dav3r committed:
Remove a now-unused import

dav3r committed:
Add hostname to top risky hosts dataframe

dav3r committed:
Include hostname column when calculating vulnerability deltas

This is needed so that hostnames can be reported in various tables in the report appendices.

dav3r committed:
Add hostnames to "top high-risk hosts" table

dav3r committed:
Add hostnames to "mitigated vulnerabilities" table in appendix

dav3r committed:
Add hostnames to "new vulnerabilities detected" table in appendix

dav3r committed:
Add hostnames to "re-detected vulnerabilities" table in appendix

dav3r committed:
Add hostnames to "recently-detected vulnerabilities" table in appendix

dav3r committed:
Fix a bug that could show duplicate ports in the report

This is in the "Critical and High Vulnerability Mitigations by IP Address" appendix.

dav3r committed:
Reduce hostname column width in some appendices

Also, remove an unnecessary character in some table column specifications.

dav3r committed:
Add hostname column to "Critical and High Vulnerability Mitigations by IP Address" appendix table

dav3r committed:
Add hostname column to False Positive appendix tables

dav3r committed:
Include hostname column when creating recently_detected_closed_tickets dataframe

This was overlooked previously.

dav3r committed:
Include hostnames in detailed findings appendix table

dav3r committed:
Add a new attachment showing the hostnames that were in scope for the report

Also, rename the previous "scope.csv" attachment to "scope-ip.csv" to reduce confusion between the two scope attachments.

dav3r committed:
Improve output of affected hosts in "Detailed Findings" appendix

This change makes the output more readable by combining all IPs that resolve to the same hostname, e.g. host.sample.gov (1.1.1.1, 2.2.2.2, 3.3.3.3, etc.).  Previously, it was displaying as host.sample.gov (1.1.1.1), host.sample.gov (2.2.2.2), host.sample.gov (3.3.3.3), etc.

dav3r committed:
Fix a bug that was creating unintended output in the detailed findings appendix table

dav3r committed:
Exclude hostname column from services attachment if there are no hostnames in the data

dav3r committed:
Ensure CybEx scorecard query includes hosts owned by default CYHY entity whose hostnames are

This means the query will correctly include hosts with hostnames assigned to the relevant entities, even if CYHY owns the actual HostDoc(s).

jsf9k reviewed (approved):
Approved _harder!!!_

dav3r committed:
Align versions of cyhy-reports package, cyhy-report script, and cyhy-cybex-scorecard script to match previously-updated version of cyhy-notification script

All are now v1.1.0.

dav3r committed:
Use correct syntax to define columns

Thanks @copilot for bringing this to my attention, even if your suggested fix wasn't quite right.

dav3r committed:
Change wonky-looking column from left-justification to right-justification in Top Hosts by Weighted Risk table

dav3r committed:
Add column padding to a bunch of tables throughout the customer report

This makes them much easier to read, since the columns are no longer all smushed together.

dav3r commented on cyhy_report/customer/report.mustache:
```diff
@@ -1504,12 +1504,12 @@ This section lists the new \glspl{vulnerability} that were detected for the firs
+	<<#display_owner>> & <</display_owner>>& <<#has_hostnames_in_tix>>& <</has_hostnames_in_tix>>& Address & & (UTC) & (UTC) & Days \\
 	\midrule \endhead
-	<<#redetected_vulnerabilities>><<#display_owner>><<&owner>> & <</display_owner>><<&plugin_name>> & <<&severity>> & <<&ip>> & <<&port>> & \isodate\formatdate<<&time_opened_date_tex>> \  \formattime<<&time_opened_time_tex>> & \isodate\formatdate<<&last_detected_date_tex>> \ \formattime<<&last_detected_time_tex>> & <<&age>> \\
+	<<#redetected_vulnerabilities>><<#display_owner>><<&owner>> & <</display_owner>><<&plugin_name>> & <<&severity>> <<#has_hostnames_in_tix>><<&hostname>> & <</has_hostnames_in_tix>>& <<&ip>> & <<&port>> & \isodate\formatdate<<&time_opened_date_tex>> \  \formattime<<&time_opened_time_tex>> & \isodate\formatdate<<&last_detected_date_tex>> \ \formattime<<&last_detected_time_tex>> & <<&age>> \\
```
You definitely found a problem, but the suggested fix is incorrect.  Please see https://github.com/cisagov/cyhy-reports/pull/128/commits/a44a83b17d9143af3521fceda69c5f6bd1ab667d for the correct fix.

dav3r commented on cyhy_report/customer/report.mustache:
```diff
@@ -1520,12 +1520,12 @@ This section lists the \glspl{vulnerability} that were previously detected, then
+	<<#display_owner>> & <</display_owner>> & <<#has_hostnames_in_tix>>& <</has_hostnames_in_tix>>& Address &  & (UTC) & (UTC) & Days \\
 	\midrule \endhead
-	<<#recently_detected_closed_tickets>><<#display_owner>><<&owner>> & <</display_owner>><<&name>> & <<&severity>> & <<&ip>> & <<&port>> & \isodate\formatdate<<&time_opened_date_tex>> \  \formattime<<&time_opened_time_tex>> & \isodate\formatdate<<&last_detected_date_tex>> \ \formattime<<&last_detected_time_tex>> & <<&age>> \\
+	<<#recently_detected_closed_tickets>><<#display_owner>><<&owner>> & <</display_owner>><<&name>> & <<&severity>><<#has_hostnames_in_tix>><<&hostname>> & <</has_hostnames_in_tix>>& <<&ip>> & <<&port>> & \isodate\formatdate<<&time_opened_date_tex>> \  \formattime<<&time_opened_time_tex>> & \isodate\formatdate<<&last_detected_date_tex>> \ \formattime<<&last_detected_time_tex>> & <<&age>> \\
```
You definitely found a problem, but the suggested fix is incorrect.  Please see https://github.com/cisagov/cyhy-reports/pull/128/commits/a44a83b17d9143af3521fceda69c5f6bd1ab667d for the correct fix.

dav3r committed:
Remove unnecessary whitespace

dav3r commented on cyhy_report/customer/generate_report.py:
```diff
@@ -2869,17 +2862,30 @@ def __generate_risky_services_attachment(self):
+            "ip_int",
+            "ip",
+            "name",
+        ]
```
Good catch, fixed in https://github.com/cisagov/cyhy-reports/pull/128/commits/f3af739941f074313485266bcda151658a3146d5.

jsf9k reviewed (approved):
_APPROVALLLLLLL!!!!!!_

felddy reviewed (approved):
`v1.1.0`‼️
[Hachi Machi!](https://www.youtube.com/watch?v=O4foeo3oY-E)
This is pretty amazing.  Thanks for cleaning up the formatting as you went.
No notes! 💪
